core: improve & refactor rpc protocol

tortmayr opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does
Implements the tweaks and fixes for the rpc system discussed in https://github.com/eclipse-theia/theia/issues/12368
- Ensure that pending requests are properly rejected when the underlying service channel is closed.
- Remove obsolete id-property from `NotificationMessage`s. Ids are only required fro matching request-response pairs.
- Use a deferred in the `RpcProxyFactory` for protocol initialization

Contributed on behalf of STMicroelectronics
<!-- Include relevant issues and describe how they are addressed. -->

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
Nothing to test in particular. The change should not break any functionality. Everything should work as before
#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

Fixes #12581
#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

tsmaeder commented on packages/core/src/node/logger-backend-module.ts:
```diff
@@ -54,11 +54,11 @@ export function bindLogger(bind: interfaces.Bind, props?: {
-                setRootLogger(ctx.container.get<ILogger>(ILogger));
-            }
-        }));
+    ({
```
While we're fixing the formatting, why not do it right?
```
bind(BackendApplicationContribution).toDynamicValue(ctx =>{
    initialize(): void {
        setRootLogger(ctx.container.get<ILogger>(ILogger));
    }
});
```

tortmayr replied:
That does not work and gives a. You cannot directly return object without surrounding it with (). Otherwise the compiler will treat it like a function block and throw an error.

tsmaeder reviewed (commented):
A couple of comments inline. In general, it might have been easier to make the renames a separate, non-functional PR. Right now the renames hide the functional changes a bit.

tortmayr commented:
> In general, it might have been easier to make the renames a separate, non-functional PR. Right now the renames hide the functional changes a bit.
Yeah that a good point. I have addressed your review and  pushed an update that removes the renames and only contains the functional changes.

I will open a separate PR for the renaming changes.

tortmayr commented on packages/core/src/common/messaging/proxy-factory.ts:
```diff
@@ -76,87 +88,87 @@ const defaultRPCConnectionFactory: RpcConnectionFactory = (channel, requestHandl
-        );
-        this.connectionPromise.then(connection => {
-            connection.channel.onClose(() => {
+        this.rpcDeferred.promise.then(protocol => {
```
Fixed

tortmayr commented on packages/core/src/common/messaging/proxy-factory.ts:
```diff
@@ -54,20 +56,30 @@ export class JsonRpcConnectionHandler<T extends object> implements ConnectionHan
-const defaultRPCConnectionFactory: RpcConnectionFactory = (channel, requestHandler) => new RpcProtocol(channel, requestHandler);
+const defaultRpcProtocolFactory: RpcProtocolFactory = (channel, requestHandler) => new RpcProtocol(channel, requestHandler);
+
+export interface RpcProxyFactoryOptions {
```
👍🏼

tortmayr commented on packages/core/src/common/messaging/proxy-factory.ts:
```diff
@@ -76,87 +88,87 @@ const defaultRPCConnectionFactory: RpcConnectionFactory = (channel, requestHandl
                 this.waitForConnection();
             });
+            protocol.channel.onError(err => {
+                if (this.rpcDeferred.state !== 'resolved') {
```
Removed the lines in question. They were committed on accident (rebasing shenanigans)

tortmayr commented on packages/core/src/common/message-rpc/rpc-message-encoder.ts:
```diff
@@ -111,7 +110,7 @@ export interface RpcMessageDecoder {
     cancel(buf: WriteBuffer, requestId: number): void;
 
-    notification(buf: WriteBuffer, requestId: number, method: string, args: any[]): void
+    notification(buf: WriteBuffer, method: string, args: any[]): void
```
Good point. I opted for making the notification id optional. This way we can still use it for tracing with the base protocol.
Adopters that use a custom protocol or reuse the RpcMessageEncoder can simply omit the id if they don't need.

tsmaeder commented:
@tortmayr I checked out the latest state of the branch and for me, it is quite broken: plugins do not start at all. There is nothing obviously wrong when looking at the logs.

tsmaeder commented:
@tortmayr something might be wrong with my system. Hold on before you start panicking ;-)

tsmaeder reviewed (approved):
Looks good, now.
